### PR TITLE
Allow to configure precheck DNS nameservers and CA certificates for private ACME providers

### PR DESCRIPTION
**Pull request by MartinWeindel**
**How to categorize this PR?**

/area  control-plane
/kind enhancement
/priority normal

**What this PR does / why we need it**:
If a custom certificate authority should be used in an on-premise environment, requesting certificates via ACME may
need additional configuration.
1. By default, the `cert-controller-manager` uses public DNS servers for prechecking the availability of the DNS challenge records. These servers can now be overwritten in the controller registration providerConfig with `values.certificateConfig.precheckNameservers`
2. The HTTPS ACME endpoint may use a self-signed certificate. This can be made known to the  `cert-controller-manager` with providing a custom list of CA certificates (`values.certificateConfig.caCertificates` in the providerConfig)

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Allow to configure precheck DNS nameservers and CA certificates for private ACME providers
```

### CODE REVIEW

File: pkg/apis/config/v1alpha1/types.go [hunk 1]
```diff
@@ -55,3 +58,13 @@ type ACME struct {
 	// +optional
 	PropagationTimeout *metav1.Duration `json:"propagationTimeout,omitempty"`
 }
+
+// Custom contains optional custom configuration for the cert-controller-manager deployment
+type Custom struct {
```
Comment (timuthy) [inline pkg/apis/config/v1alpha1/types.go:63]: What do you think about reusing the `ACME` struct here? `Custom` sounds very generic and the fields below are actually ACME related.

File: pkg/apis/config/v1alpha1/types.go [hunk 2]
```diff
@@ -55,3 +58,13 @@ type ACME struct {
...
+	// PrecheckNameservers is the value for the `--issuer.precheck-nameservers` command line parameter
+	// +optional
+	PrecheckNameservers *string `json:"precheckNameservers,omitempty"`
```
Comment (timuthy) [inline pkg/apis/config/v1alpha1/types.go:66]: Can you add an input validation? I suppose only conformant IP addresses are allowed here.

File: charts/gardener-extension-shoot-cert-service/templates/_helpers.tpl [hunk 3]
```diff
@@ -21,6 +21,15 @@ acme:
   privateKey: |
 {{ .Values.certificateConfig.defaultIssuer.acme.privateKey | trim | indent 4 }}
   {{- end }}
+{{- if .Values.customConfig }}
+custom:
```
Comment (timuthy) [inline charts/gardener-extension-shoot-cert-service/templates/_helpers.tpl:25]: What do you think about reusing the `ACME` structure here? `custom` sounds very generic and the fields below are actually ACME related.

File: charts/gardener-extension-shoot-cert-service/values.yaml [hunk 4]
```diff
@@ -40,6 +40,18 @@ certificateConfig:
     #   ...
     #   -----END RSA PRIVATE KEY-----
 
+customConfig:
```
Comment (timuthy) [inline charts/gardener-extension-shoot-cert-service/values.yaml:43]: See previous comment.

File: pkg/apis/config/v1alpha1/types.go [hunk 5]
```diff
@@ -55,3 +58,13 @@ type ACME struct {
...
+	// CACertificates are custom root certificates to be made available for the cert-controller-manager
+	// +optional
+	CACertificates *string `json:"caCertificates,omitempty"`
```
Comment (timuthy) [inline pkg/apis/config/v1alpha1/types.go:69]: Same here. Shall we validate the input? I know it's also not done for `PrivateKey` but probably good to fail early.

**Review by timuthy — CHANGES_REQUESTED** [review]
Logic wise it looks good to me. I have a few suggestions, generally proposing a more self expressive name then `customConfig` which is kind of vague.

File: pkg/apis/config/v1alpha1/types.go (hunk 1 above)
Comment (MartinWeindel) [reply pkg/apis/config/v1alpha1/types.go]: `ACME`struct is reused as proposed.

File: pkg/apis/config/v1alpha1/types.go (hunk 2 above)
Comment (MartinWeindel) [reply pkg/apis/config/v1alpha1/types.go:60]: validation added

File: pkg/apis/config/v1alpha1/types.go (hunk 5 above)
Comment (MartinWeindel) [reply pkg/apis/config/v1alpha1/types.go:63]: I added some basic validation for correct PEM format

**Review by timuthy — APPROVED** [review]
/lgtm

**Pull request merged**